### ISSUE TITLE
BatchedPreparedStatement doesn't work if statement is re-used

### ISSUE THREAD

**Issue by AnghelLeonard** [jOOQ/jOOQ#11443]
### Expected behavior
A new batch to be created.

### Actual behavior
Batch `null`.

### Steps to reproduce the problem
This code (notice the batch size of 2):

```Java
try ( BatchedConnection conn = new BatchedConnection(DriverManager.getConnection(
                "jdbc:mysql://localhost:3306/classicmodels", "root", "root"), 2)) {
            
            PreparedStatement stmt = conn.prepareStatement("insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) "
                    + "values (?, ?, ?);");

            // the next 2 statements will become the first batch                        
            stmt.setInt(1, 2004);
            stmt.setLong(2, 1166L);
            stmt.setDouble(3, 543.33);
            stmt.executeUpdate();
            
            stmt.setInt(1, 2005);
            stmt.setLong(2, 1370L);
            stmt.setDouble(3, 9022.20);
            stmt.executeUpdate();
            
            // reached batch limit so this is the second batch
            stmt.setInt(1, 2003);
            stmt.setLong(2, 1166L);
            stmt.setDouble(3, 3213.0);
            stmt.executeUpdate();
                        
            // since the following SQL string is different, next statements represents the third batch
            stmt = conn.prepareStatement("insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) "
                    + "values (?, ?, ?, ?);");
            
            stmt.setInt(1, 2004);
            stmt.setLong(2, 1166L);
            stmt.setDouble(3, 4541.35);
            stmt.setString(4, "UP");
            stmt.executeUpdate();
            
            stmt.setInt(1, 2005);
            stmt.setLong(2, 1370L);
            stmt.setDouble(3, 1282.64);
            stmt.setString(4, "DOWN");
            stmt.executeUpdate();            

        } catch (SQLException ex) {
            Logger.getLogger(ClassicModelsRepository.class.getName()).log(Level.SEVERE, null, ex);
        }
```
Log (check the bolded code):

```text
Batched 1 times: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) values (?, ?, ?);
Batched 2 times: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) values (?, ?, ?);
Executed with 2 batched items: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) values (?, ?, ?);
**Batched 1 times: null**
Batched 1 times: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) values (?, ?, ?, ?);
Batched 2 times: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) values (?, ?, ?, ?);
Executed with 2 batched items: insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) values (?, ?, ?, ?);
```

And, the record is not inserted in the database.

### Versions

- jOOQ:3.14.7
- Java:14
- Database (include vendor):MySQL
- OS:Windows 10
- JDBC Driver (include name if inofficial driver):-

**Comment by lukaseder** [comment]
Thanks a lot for your report. Interesting. When running a similar test with jOOQ created queries, the test works as expected. I wonder if this is because you're not closing the prepared statements. Will investigate right away.

**Comment by lukaseder** [comment]
Oh, I see. The problem is that you're re-using the `PreparedStatement` instance. This should work:

```java
try ( BatchedConnection conn = new BatchedConnection(DriverManager.getConnection(
                "jdbc:mysql://localhost:3306/classicmodels", "root", "root"), 2)) {
            
	String sql1 = "insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) "
			+ "values (?, ?, ?)";

	try (PreparedStatement stmt = conn.prepareStatement(sql1)) {
		stmt.setInt(1, 2004);
		stmt.setLong(2, 1166L);
		stmt.setDouble(3, 543.33);
		stmt.executeUpdate();
	}
	
	try (PreparedStatement stmt = conn.prepareStatement(sql1)) {
		stmt.setInt(1, 2005);
		stmt.setLong(2, 1370L);
		stmt.setDouble(3, 9022.20);
		stmt.executeUpdate();
	}
	
	// reached batch limit so this is the second batch
	try (PreparedStatement stmt = conn.prepareStatement(sql1)) {
		stmt.setInt(1, 2003);
		stmt.setLong(2, 1166L);
		stmt.setDouble(3, 3213.0);
		stmt.executeUpdate();
	}
				
	// since the following SQL string is different, next statements represents the third batch
	String sql2 = "insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) "
			+ "values (?, ?, ?, ?)";
	
	try (PreparedStatement stmt = conn.prepareStatement(sql2)) {
		stmt.setInt(1, 2004);
		stmt.setLong(2, 1166L);
		stmt.setDouble(3, 4541.35);
		stmt.setString(4, "UP");
		stmt.executeUpdate();
	}
	
	try (PreparedStatement stmt = conn.prepareStatement(sql2)) {
		stmt.setInt(1, 2005);
		stmt.setLong(2, 1370L);
		stmt.setDouble(3, 1282.64);
		stmt.setString(4, "DOWN");
		stmt.executeUpdate();            
	}
} catch (SQLException ex) {
	Logger.getLogger(ClassicModelsRepository.class.getName()).log(Level.SEVERE, null, ex);
}
```

Of course, your version should work as well.

**Comment by AnghelLeonard** [comment]
Ok, I'll do it this way. Just for the record, I tried this as well and it does the same `null` thing:

```Java
public void batchedConnectionUsage() {

        try ( BatchedConnection conn = new BatchedConnection(DriverManager.getConnection(
                "jdbc:mysql://localhost:3306/classicmodels", "root", "root"), 2)) {

            try ( PreparedStatement stmt = conn.prepareStatement("insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`) "
                    + "values (?, ?, ?);")) {

                // the next 2 statements will become the first batch                        
                stmt.setInt(1, 2004);
                stmt.setLong(2, 1166L);
                stmt.setDouble(3, 543.33);
                stmt.executeUpdate();

                stmt.setInt(1, 2005);
                stmt.setLong(2, 1370L);
                stmt.setDouble(3, 9022.20);
                stmt.executeUpdate();

                // reached batch limit so this is the second batch
                stmt.setInt(1, 2003);
                stmt.setLong(2, 1166L);
                stmt.setDouble(3, 3213.0);
                stmt.executeUpdate();
            } 

            // since the following SQL string is different, next statements represents the third batch
            try ( PreparedStatement stmt = conn.prepareStatement("insert into `classicmodels`.`sale` (`fiscal_year`, `employee_number`, `sale`, `trend`) "
                    + "values (?, ?, ?, ?);")) {

                stmt.setInt(1, 2004);
                stmt.setLong(2, 1166L);
                stmt.setDouble(3, 4541.35);
                stmt.setString(4, "UP");
                stmt.executeUpdate();

                stmt.setInt(1, 2005);
                stmt.setLong(2, 1370L);
                stmt.setDouble(3, 1282.64);
                stmt.setString(4, "DOWN");
                stmt.executeUpdate();
            }
            
        } catch (SQLException ex) {
            Logger.getLogger(ClassicModelsRepository.class.getName()).log(Level.SEVERE, null, ex);
        }
```

**Comment by lukaseder** [comment]
Yes, the problem is re-using the statement. which obviously shouldn't be a problem.

**Comment by lukaseder** [comment]
A related, but not exactly the same case would be if we had the two statements open at the same time, and then alternate executions without closing any statement in between.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.8 (#11445). Thanks again for your report!
